src/main/java/io/reactivex/internal/operators/flowable/FlowableBufferBoundary.java

```diff
@@ -98,214 +119,325 @@ public void onSubscribe(Subscription s) {
+                queue.offer(buffers.remove(idx));
             }
+            if (makeDone) {
+                done = true;
+            }
+            drain();
+        }
 
-            BufferCloseSubscriber<T, U, Open, Close> bcs = new BufferCloseSubscriber<T, U, Open, Close>(b, this);
-            resources.add(bcs);
+        void boundaryError(Disposable subscriber, Throwable ex) {
+            SubscriptionHelper.cancel(upstream);
+            subscribers.delete(subscriber);
+            if (errors.addThrowable(ex)) {
+                subscribers.dispose();
+                synchronized (this) {
+                    buffers = null;
+                }
+                done = true;
+                drain();
+            } else {
+                RxJavaPlugins.onError(ex);
+            }
+        }
 
-            windows.getAndIncrement();
+        void drain() {
+            if (getAndIncrement() != 0) {
+                return;
+            }
 
-            p.subscribe(bcs);
-        }
+            int missed = 1;
+            long e = emitted;
+            Subscriber<? super C> a = actual;
+            SpscLinkedArrayQueue<C> q = queue;
+
+            for (;;) {
+                long r = requested.get();
+
+                while (e != r) {
+                    if (cancelled) {
+                        q.clear();
+                        return;
+                    }
+
+                    boolean d = done;
+                    if (d && errors.get() != null) {
+                        q.clear();
+                        Throwable ex = errors.terminate();
+                        a.onError(ex);
+                        return;
+                    }
+
+                    C v = q.poll();
+                    boolean empty = v == null;
+
+                    if (d && empty) {
+                        a.onComplete();
+                        return;
+                    }
+
+                    if (empty) {
+                        break;
+                    }
+
+                    a.onNext(v);
+                    e++;
+                }
 
-        void openFinished(Disposable d) {
-            if (resources.remove(d)) {
-                if (windows.decrementAndGet() == 0) {
-                    complete();
+                if (e == r) {
+                    if (cancelled) {
+                        q.clear();
+                        return;
+                    }
+
+                    if (done) {
+                        if (errors.get() != null) {
+                            q.clear();
+                            Throwable ex = errors.terminate();
+                            a.onError(ex);
+                            return;
+                        } else
+                        if (q.isEmpty()) {
+                            a.onComplete();
+                            return;
+                        }
+                    }
+                }
+
+                emitted = e;
+                missed = addAndGet(-missed);
```

Comment:
`requested` not getting reduced by `emitted`?

Reply:
No longer necessary in these type of drain loops. `requested` can only grow and the `emitted` will follow it separately. This saves an atomic decrement here.

Reply:
ok, beaut